### PR TITLE
Message feature update

### LINKED ISSUE #84: Finish Message Component
### What is this for?

website

### Feature

As a user, I want the message component to have parity with Discord messages so I can get all the information I need

### Ideal solution or implementation

- [x] Markdown Support
- [ ] #318
- [x] Monospace codeblock font
- [x] Opening images as modals
- [x] Code Blocks
- [x] Links to open in Discord & copy URL
- [x] Revisit Ignored Unescaped HTML

### Alternative solutions or implementations

_No response_

### Other context

_No response_

### PR DESCRIPTION
**Pull request by JoltCode**
# Description

Adds the following message features:
- Modal for image attachments

Fixes #84 (issue)

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?

N/A - UI

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works 
- [x] I have updated .env.example if I added a new environment variable
- [x] My PR title follows the [semantic commits](https://www.conventionalcommits.org/en/v1.0.0/) style

### CODE REVIEW

File: packages/ui/src/components/primitives/Message.tsx [hunk 1]
```diff
@@ -24,15 +27,74 @@ export function useMessageContext() {
...
+
+					<Button
+						className="p-1"
+						onClick={() =>
+							copyToClipboard(
+								`https://discord.com/${message.serverId}/${message.id}`,
+							)
+						}
+					>
+						<CopyIcon />
```
Comment (RhysSullivan) [inline packages/ui/src/components/primitives/Message.tsx:94]: Copy URL isn't needed, if we were to have it the place that would make sense for it is in the headline for the message result page

File: packages/ui/src/components/primitives/Message.tsx [hunk 2]
```diff
@@ -24,15 +27,74 @@ export function useMessageContext() {
...
 			<div className="flex w-full flex-row items-center gap-2 font-body text-lg text-black/[.7] dark:text-white/[.47]">
 				<DiscordAvatar user={message.author} size="sm" />
 				<span className="mr-1">{message.author.name}</span>
-				<span className="ml-auto">{getSnowflakeUTCDate(message.id)}</span>
+				<div className="mr-4 ml-auto flex flex-row gap-2">
+					<LinkButton
+						href={`https://discord.com/${message.serverId}/${message.id}`}
+						className="p-1"
+					>
+						<ExternalLinkIcon />
```
Comment (RhysSullivan) [inline packages/ui/src/components/primitives/Message.tsx:83]: Change to a discord icon

File: packages/ui/src/components/primitives/Message.tsx [hunk 3]
```diff
@@ -24,15 +27,74 @@ export function useMessageContext() {
...
 	return (
 		<div className="flex w-full min-w-0 gap-2">
 			{/* TODO: sort out responsive styling */}
 			<div className="flex w-full flex-row items-center gap-2 font-body text-lg text-black/[.7] dark:text-white/[.47]">
 				<DiscordAvatar user={message.author} size="sm" />
 				<span className="mr-1">{message.author.name}</span>
-				<span className="ml-auto">{getSnowflakeUTCDate(message.id)}</span>
+				<div className="mr-4 ml-auto flex flex-row gap-2">
+					<LinkButton
+						href={`https://discord.com/${message.serverId}/${message.id}`}
```
Comment (RhysSullivan) [inline packages/ui/src/components/primitives/Message.tsx:80]: Incorrrect messages link, see other comment on this

**Review by RhysSullivan — CHANGES_REQUESTED** [review]
Good start, listed a few requested changes

**Comment by RhysSullivan** [comment]
Extra tasks:

- For now, only display attachments with .png and .mp4 (or handle other types better, atm they 404)
- Only show Discord icon on hover on desktop
- On the message result page, only show the discord icon on the first message

File: packages/ui/src/components/pages/MessageResultPage.tsx [hunk 4]
```diff
@@ -121,9 +123,10 @@ export function MessageResultPage({
 			</div>
 			<div className="rounded-md">
 				<h1 className="mb-4 rounded-sm border-b-2 border-solid border-neutral-400 pb-2 text-3xl dark:border-neutral-600 dark:text-white">
-					{thread ? thread.name : channel.name}
+					{/* TODO: Add proper icon for a thread/channel */}
+					{thread ? `#${thread.name}` : `#${channel.name}`}
```
[the remark this answers is not in the document]
Comment (JoltCode) [reply packages/ui/src/components/pages/MessageResultPage.tsx:127]: Where is this?

File: packages/ui/src/components/primitives/Message.tsx (hunk 2 above)
Comment (JoltCode) [reply packages/ui/src/components/primitives/Message.tsx:83]: Done

File: packages/ui/src/components/primitives/Message.tsx [hunk 5]
```diff
@@ -24,15 +27,74 @@ export function useMessageContext() {
...
+						className="p-1"
+					>
+						<ExternalLinkIcon />
+					</LinkButton>
+
+					<Button
+						className="p-1"
+						onClick={() =>
+							copyToClipboard(
+								`https://discord.com/${message.serverId}/${message.id}`,
```
[the remark this answers is not in the document]
Comment (JoltCode) [reply packages/ui/src/components/primitives/Message.tsx:90]: Copy to clip removed for now

File: packages/ui/src/components/primitives/Message.tsx [hunk 6]
```diff
@@ -24,15 +27,74 @@ export function useMessageContext() {
 	return context;
 }
 
+const CopyIcon = () => {
```
[the remark this answers is not in the document]
Comment (JoltCode) [reply packages/ui/src/components/primitives/Message.tsx:30]: Done

File: packages/ui/src/components/primitives/Message.tsx [hunk 7]
```diff
@@ -24,15 +27,74 @@ export function useMessageContext() {
...
+			<path
+				strokeLinecap="round"
+				strokeLinejoin="round"
+				d="M15.666 3.888A2.25 2.25 0 0013.5 2.25h-3c-1.03 0-1.9.693-2.166 1.638m7.332 0c.055.194.084.4.084.612v0a.75.75 0 01-.75.75H9a.75.75 0 01-.75-.75v0c0-.212.03-.418.084-.612m7.332 0c.646.049 1.288.11 1.927.184 1.1.128 1.907 1.077 1.907 2.185V19.5a2.25 2.25 0 01-2.25 2.25H6.75A2.25 2.25 0 014.5 19.5V6.257c0-1.108.806-2.057 1.907-2.185a48.208 48.208 0 011.927-.184"
+			/>
+		</svg>
+	);
+};
+
+const ExternalLinkIcon = () => {
```
[the remark this answers is not in the document]
Comment (JoltCode) [reply packages/ui/src/components/primitives/Message.tsx:49]: Done

File: packages/ui/src/components/primitives/Message.tsx [hunk 8]
```diff
@@ -50,6 +112,55 @@ export const MessageContents = () => {
 	);
 };
 
+const CloseIcon = () => {
```
[the remark this answers is not in the document]
Comment (JoltCode) [reply packages/ui/src/components/primitives/Message.tsx:115]: Done

File: packages/ui/src/components/primitives/Message.tsx [hunk 9]
```diff
@@ -50,6 +112,55 @@ export const MessageContents = () => {
...
+			<path
+				strokeLinecap="round"
+				strokeLinejoin="round"
+				d="M6 18L18 6M6 6l12 12"
+			/>
+		</svg>
+	);
+};
+
+export const MessageModalWrapper = ({
```
[the remark this answers is not in the document]
Comment (JoltCode) [reply packages/ui/src/components/primitives/Message.tsx:134]: Not anymore :+1:

File: packages/ui/src/components/primitives/Message.tsx [hunk 10]
```diff
@@ -50,6 +112,55 @@ export const MessageContents = () => {
...
+			<AlertDialog.Trigger asChild>{children}</AlertDialog.Trigger>
+			<AlertDialog.Portal>
+				<AlertDialog.Overlay className="fixed inset-0 z-50 flex h-full w-full items-center justify-center bg-black/75" />
+				<AlertDialog.Content className="fixed top-0 left-0 z-[75] flex h-full w-full flex-col items-center justify-center p-4">
+					<div className="absolute top-0 right-0 z-[100] p-2">
+						<AlertDialog.Cancel className="flex h-8 w-8 items-center justify-center rounded-full bg-white transition-colors duration-200 focus:outline-none focus:ring focus:ring-gray-300 hover:bg-gray-200 dark:bg-black dark:hover:bg-gray-800">
+							<CloseIcon />
+						</AlertDialog.Cancel>
+					</div>
+					<img
```
[the remark this answers is not in the document]
Comment (JoltCode) [reply packages/ui/src/components/primitives/Message.tsx:153]: Yeah, h/w is optional in the attachment, I think we should leave it as is for now, then in a future PR we can sort out all the images which don't have a required h/w (we can use the fill property for non avail h/w)

File: packages/ui/src/components/primitives/Message.tsx [hunk 11]
```diff
@@ -66,16 +177,18 @@ export const MessageAttachments = () => {
...
-					src={attachment.url}
-					style={{
-						width: 'fit-content',
-						height: 'auto',
-						objectFit: 'cover',
-					}}
-					alt={attachment.description ? attachment.description : 'Image'}
-				/>
+				<MessageModalWrapper attachment={attachment}>
+					<img
```
[the remark this answers is not in the document]
Comment (JoltCode) [reply packages/ui/src/components/primitives/Message.tsx:181]: Same reason as other non next/image

File: packages/ui/src/components/primitives/Message.tsx (hunk 3 above)
Comment (JoltCode) [reply packages/ui/src/components/primitives/Message.tsx:80]: Should be fixed - please check if this is okay now!

File: packages/ui/src/components/primitives/Message.tsx (hunk 1 above)
Comment (JoltCode) [reply packages/ui/src/components/primitives/Message.tsx:94]: Removed!

File: packages/ui/src/components/primitives/Message.tsx
[the remark this answers is not in the document]
Comment (JoltCode) [reply packages/ui/src/components/primitives/Message.tsx:1]: Which states? Perhaps we could look at just moving this to another issue? We want to make sure we keep the PR's small - along with getting this merged asap :+1:

File: apps/docs/src/pages/_app.tsx [hunk 12]
```diff
@@ -17,9 +17,6 @@ const MyApp: AppType<{ session: Session | null }> = ({
 	pageProps: { session, ...pageProps },
 }) => {
 	useEffect(() => {
-		hljs.configure({
-			ignoreUnescapedHTML: true,
```
[the remark this answers is not in the document]
Comment (JoltCode) [reply apps/docs/src/pages/_app.tsx:21]: I can't see any noticeable warnings in the logs on the docs home page (using dev mode).

**Pull request merged**